File: sdk/go/events.go
package fathom

import (
	"crypto/rand"
	"fmt"
	"runtime"
	"strings"
	"time"
)

// SdkEvent represents a single event to be sent to FATHOM
// All fields use camelCase to match the server wire format exactly
type SdkEvent struct {
	Type        string                 `json:"type"`
	Title       string                 `json:"title"`
	Severity    string                 `json:"severity"`
	Message     string                 `json:"message,omitempty"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
	Timestamp   string                 `json:"timestamp"`
	SessionID   string                 `json:"sessionId"`
	AppName     string                 `json:"appName"`
	Environment string                 `json:"environment"`
	Tags        []string               `json:"tags,omitempty"`
}

// generateSessionID generates a session ID in format <millis>-<8-hex>
func generateSessionID() string {
	millis := time.Now().UnixMilli()

	// Generate 4 random bytes = 8 hex characters
	b := make([]byte, 4)
	if _, err := rand.Read(b); err != nil {
		// Fallback if random fails (shouldn't happen in practice)
		return fmt.Sprintf("%d-00000000", millis)
	}

	hexPart := fmt.Sprintf("%08x", b)
	return fmt.Sprintf("%d-%s", millis, hexPart)
}

// generateErrorTitle creates a title from an error or title string
func generateErrorTitle(err error) string {
	if err == nil {
		return "Error"
	}

	msg := err.Error()
	if len(msg) > 80 {
		msg = msg[:80]
	}
	return fmt.Sprintf("error: %s", msg)
}

// APIEvent creates an API tracking event with auto-severity mapping
func APIEvent(method, endpoint string, statusCode int, durationMs int64, metadata map[string]interface{}) *SdkEvent {
	if metadata == nil {
		metadata = make(map[string]interface{})
	}

	// Auto-map severity based on status code
	var severity string
	if statusCode >= 500 {
		severity = "error"
	} else if statusCode >= 400 {
		severity = "warning"
	} else {
		severity = "info"
	}

	// Build metadata
	fullMetadata := make(map[string]interface{})
	for k, v := range metadata {
		fullMetadata[k] = v
	}
	fullMetadata["method"] = method
	fullMetadata["endpoint"] = endpoint
	fullMetadata["statusCode"] = statusCode
	fullMetadata["durationMs"] = durationMs

	return &SdkEvent{
		Type:        "api",
		Title:       fmt.Sprintf("%s %s", method, endpoint),
		Severity:    severity,
		Message:     "",
		Metadata:    fullMetadata,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
}

// ErrorEvent creates an error tracking event from an error
func ErrorEvent(err error, metadata map[string]interface{}) *SdkEvent {
	if metadata == nil {
		metadata = make(map[string]interface{})
	}

	errorMetadata := make(map[string]interface{})
	for k, v := range metadata {
		errorMetadata[k] = v
	}

	title := "Error"
	if err != nil {
		errType := fmt.Sprintf("%T", err)
		// Strip package name to get just the type name
		if idx := strings.LastIndex(errType, "."); idx >= 0 {
			errType = errType[idx+1:]
		}
		errMsg := err.Error()

		errorMetadata["name"] = errType
		errorMetadata["message"] = errMsg

		// Capture stack trace
		buf := make([]byte, 4096)
		n := runtime.Stack(buf, false)
		if n > 0 {
			errorMetadata["stack"] = string(buf[:n])
		}

		// Title format: "error: <msg first 80 chars>"
		if len(errMsg) > 80 {
			errMsg = errMsg[:80]
		}
		title = fmt.Sprintf("error: %s", errMsg)
	}

	return &SdkEvent{
		Type:        "error",
		Title:       title,
		Severity:    "error",
		Message:     "",
		Metadata:    errorMetadata,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
}

// AuditEvent creates an audit tracking event
func AuditEvent(title string, message string, severity string, metadata map[string]interface{}) *SdkEvent {
	if metadata == nil {
		metadata = make(map[string]interface{})
	}

	fullMetadata := make(map[string]interface{})
	for k, v := range metadata {
		fullMetadata[k] = v
	}
	if message != "" {
		fullMetadata["message"] = message
	}

	if severity == "" {
		severity = "info"
	}

	return &SdkEvent{
		Type:        "audit",
		Title:       title,
		Severity:    severity,
		Message:     "",
		Metadata:    fullMetadata,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
}

// MetricEvent creates a metric tracking event
func MetricEvent(title string, value float64, unit string, metadata map[string]interface{}) *SdkEvent {
	if metadata == nil {
		metadata = make(map[string]interface{})
	}

	fullMetadata := make(map[string]interface{})
	for k, v := range metadata {
		fullMetadata[k] = v
	}
	fullMetadata["value"] = value
	if unit != "" {
		fullMetadata["unit"] = unit
	}

	return &SdkEvent{
		Type:        "metric",
		Title:       title,
		Severity:    "info",
		Message:     "",
		Metadata:    fullMetadata,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
}

// CustomEvent creates a custom tracking event
func CustomEvent(eventType, title, severity string, metadata map[string]interface{}, tags []string) *SdkEvent {
	if metadata == nil {
		metadata = make(map[string]interface{})
	}

	fullMetadata := make(map[string]interface{})
	for k, v := range metadata {
		fullMetadata[k] = v
	}
	if len(tags) > 0 {
		fullMetadata["tags"] = tags
	}

	return &SdkEvent{
		Type:        eventType,
		Title:       title,
		Severity:    severity,
		Message:     "",
		Metadata:    fullMetadata,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
		Tags:        tags,
	}
}

File: sdk/python/src/fathom_sdk/events.py
"""Event dataclasses for Fathom SDK."""

from dataclasses import dataclass, field, asdict
from typing import Optional, Dict, Any, List
from datetime import datetime
import json


@dataclass
class SdkEvent:
    """Represents a single SDK event.

    All fields match the backend wire format exactly (no snake_case conversion).
    Timestamps are ISO 8601 strings, set at event-creation time.
    """

    type: str
    title: str
    severity: str
    message: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)
    timestamp: Optional[str] = None
    sessionId: Optional[str] = None
    appName: Optional[str] = None
    environment: Optional[str] = None
    tags: Optional[List[str]] = None

    def __post_init__(self):
        """Validate event fields and set defaults."""
        if not self.type:
            raise ValueError("Event 'type' is required")
        if not self.title:
            raise ValueError("Event 'title' is required")
        if not self.severity:
            raise ValueError("Event 'severity' is required")

        valid_severities = {"debug", "info", "warning", "error", "critical"}
        if self.severity not in valid_severities:
            raise ValueError(
                f"Event 'severity' must be one of {valid_severities}, got {self.severity}"
            )

        # Timestamp defaults to now if not provided
        if self.timestamp is None:
            self.timestamp = datetime.utcnow().isoformat() + "Z"

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization (removing None values)."""
        result = {}
        for key, value in asdict(self).items():
            if value is not None:
                result[key] = value
        return result


def api_event(
    method: str,
    endpoint: str,
    status_code: int,
    duration_ms: int,
    *,
    metadata: Optional[Dict[str, Any]] = None,
) -> SdkEvent:
    """Create an API event with auto-severity mapping.

    Severity is set based on status code:
    - 500+: error
    - 400+: warning
    - else: info
    """
    if metadata is None:
        metadata = {}

    # Auto-map severity based on status code
    if status_code >= 500:
        severity = "error"
    elif status_code >= 400:
        severity = "warning"
    else:
        severity = "info"

    return SdkEvent(
        type="api",
        title=f"{method.upper()} {endpoint}",
        severity=severity,
        metadata={
            "method": method,
            "endpoint": endpoint,
            "statusCode": status_code,
            "durationMs": duration_ms,
            **metadata,
        },
    )


def error_event(
    exc_or_title: Optional[Exception | str] = None,
    *,
    title: Optional[str] = None,
    severity: str = "error",
    metadata: Optional[Dict[str, Any]] = None,
) -> SdkEvent:
    """Create an error event from an Exception or string.

    If an Exception is provided, captures:
    - name: exception class name
    - message: exception message
    - stack: full traceback
    """
    if metadata is None:
        metadata = {}

    error_metadata = {}

    if isinstance(exc_or_title, Exception):
        import traceback

        exc_name = type(exc_or_title).__name__
        exc_message = str(exc_or_title)
        error_metadata = {
            "name": exc_name,
            "message": exc_message,
            "stack": traceback.format_exc(),
        }
        # Title includes exception name for better visibility
        final_title = title or f"{exc_name}: {exc_message}" if exc_message else exc_name
    elif isinstance(exc_or_title, str):
        error_metadata = {"message": exc_or_title}
        final_title = title or exc_or_title
    else:
        final_title = title or "Error"

    return SdkEvent(
        type="error",
        title=final_title,
        severity=severity,
        metadata={**error_metadata, **metadata},
    )


def audit_event(
    title: str,
    *,
    message: Optional[str] = None,
    severity: str = "info",
    metadata: Optional[Dict[str, Any]] = None,
) -> SdkEvent:
    """Create an audit event."""
    if metadata is None:
        metadata = {}

    event_metadata = {}
    if message is not None:
        event_metadata["message"] = message

    return SdkEvent(
        type="audit",
        title=title,
        severity=severity,
        metadata={**event_metadata, **metadata},
    )


def metric_event(
    title: str,
    *,
    value: float | int,
    unit: str = "",
    metadata: Optional[Dict[str, Any]] = None,
) -> SdkEvent:
    """Create a metric event with value and unit."""
    if metadata is None:
        metadata = {}

    return SdkEvent(
        type="metric",
        title=title,
        severity="info",
        metadata={
            "value": value,
            "unit": unit,
            **metadata,
        },
    )
